packages/liveblocks-react/src/factory.tsx

```diff
@@ -238,22 +235,64 @@ export function createRoomContext<
   TRoomEvent,
   TThreadMetadata
 > {
-  const RoomContext = React.createContext<Room<
-    TPresence,
-    TStorage,
-    TUserMeta,
-    TRoomEvent
-  > | null>(null);
+  type TRoom = Room<TPresence, TStorage, TUserMeta, TRoomEvent>;
+
+  const RoomContext = React.createContext<TRoom | null>(null);
+
+  const roomCache = new Map<string, { room: TRoom; leave: () => void }>();
+
+  /**
+   * A cached version of .enterRoom(), where we cache each room/leave pair on
+   * a per-RoomProvider instance basis.
+   */
+  function stableEnterRoom(
```

Comment:
@nvie Do you mind explaining the significance of `roomCache` in more detail and how this functions differently from the `roomsById` variable inside [client.ts](https://github.com/liveblocks/liveblocks/pull/1226/files#diff-f7fafe86f1919de5119b3b0c680a8c12d5cd72b72c39cab803147a64faceb1e2) file?

Reply:
@nimeshnayaju I hope my explanation in [this comment](https://github.com/liveblocks/liveblocks/pull/1226#discussion_r1363332040) makes this easier to follow? I've also attempted to explain the rationale in the comment block in a2978a2d20297e9cbbb3066ba8609dd41f235c06.